src/validation.h

```diff
@@ -187,12 +187,34 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
 /** Prune block files up to a given height */
 void PruneBlockFilesManual(int nManualPruneHeight);
 
-/** (try to) add transaction to memory pool
- * plTxnReplaced will be appended to with all transactions replaced from mempool
- * @param[out] fee_out optional argument to return tx fee to the caller **/
-bool AcceptToMemoryPool(CTxMemPool& pool, TxValidationState &state, const CTransactionRef &tx,
-                        std::list<CTransactionRef>* plTxnReplaced,
-                        bool bypass_limits, bool test_accept=false, CAmount* fee_out=nullptr) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+/**
+* Validation result for a single transaction mempool acceptance.
+* When m_accepted = true, m_replaced_transactions contains a list
+* of replaced transactions and m_base_fees contains the tx fees.
+*/
+struct MempoolAcceptResult {
+    bool m_accepted;
+    TxValidationState m_state;
+
+    // Valid when m_accepted = true
+    std::list<CTransactionRef> m_replaced_transactions;
+    CAmount m_base_fees;
```

Comment:
Unless I am mistaken, incorrectly reading this will result in an uninitialized read, which is only detected by valgrind. Wouldn't it be better to make this safe for non-valgrind use via an std::optional at least. Or is that also a bit too verbose (like the std::variant approach)?

Reply:
Yeah good point, let me whip up the `std::optional` method real quick and get back to you on the verbosity

Reply:
Please see [this branch](https://github.com/glozow/bitcoin/commit/408a32b5c70a6c2d657304ae0a874dadb635f4fa), it's not too bad verbosity-wise, I have a couple ternary operators like this:

```c
const TxValidationState state = result.m_state == nullopt ? TxValidationState{} : result.m_state.value();
```

incorrect read throws a `bad optional access`

Reply:
Cool, thanks for looking. Not sure if the ternary makes sense. It seems `state` is currently assumed to be always properly initialized (redundantly to the boolean return value). So maybe it could make sense to not make this optional. I do like that the other members (fee and replaced txs) are optional in your branch.

Reply:
Ah yeah! That makes sense, leave state as-is, and optional members when they only make sense for valid transactions.